LoongArch-ELF-ABI-EN.adoc

```diff
@@ -0,0 +1,206 @@
+`assert (pop ())`
+
+|31      |R_LARCH_SOP_NOT            |Stack top operation|
+`push (!pop ())`
+
+|32      |R_LARCH_SOP_SUB            |Stack top operation|
+`opr2 = pop (), opr1 = pop (), push (opr1 - opr2)`
+
+|33      |R_LARCH_SOP_SL             |Stack top operation|
+`opr2 = pop (), opr1 = pop (), push (opr1 << opr2)`
+
+|34      |R_LARCH_SOP_SR             |Stack top operation|
+`opr2 = pop (), opr1 = pop (), push (opr1 >> opr2)`
+
+|35      |R_LARCH_SOP_ADD            |Stack top operation|
+`opr2 = pop (), opr1 = pop (), push (opr1 + opr2)`
+
+|36      |R_LARCH_SOP_AND            |Stack top operation|
+`opr2 = pop (), opr1 = pop (), push (opr1 & opr2)`
+
+|37      |R_LARCH_SOP_IF_ELSE        |Stack top operation|
+`opr3 = pop (), opr2 = pop (), opr1 = pop (), push (opr1 ? opr2 : opr3)`
+
+|38      |R_LARCH_SOP_POP_32_S_10_5  |Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [14 ... 10] = opr1 [4 ... 0]`
+ with check 5-bit signed overflow
+
+|39      |R_LARCH_SOP_POP_32_U_10_12 |Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [21 ... 10] = opr1 [11 ... 0]`
+ with check 12-bit unsigned overflow
+
+|40      |R_LARCH_SOP_POP_32_S_10_12 |Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [21 ... 10] = opr1 [11 ... 0]`
+ with check 12-bit signed overflow
+
+|41      |R_LARCH_SOP_POP_32_S_10_16 |Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [25 ... 10] = opr1 [15 ... 0]`
+ with check 16-bit signed overflow
+
+|42      |R_LARCH_SOP_POP_32_S_10_16_S2|Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [25 ... 10] = opr1 [17 ... 2]`
+ with check 18-bit signed overflow and 4-bit aligned
+
+|43      |R_LARCH_SOP_POP_32_S_5_20   |Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [24 ... 5] = opr1 [19 ... 0]`
+ with check 20-bit signed overflow
+
+|44      |R_LARCH_SOP_POP_32_S_0_5_10_16_S2|Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [4 ... 0] = opr1 [22 ... 18],`
+`(*(uint32_t *) PC) [25 ... 10] = opr1 [17 ... 2]`
+ with check 23-bit signed overflow and 4-bit aligned
+
+|45      |R_LARCH_SOP_POP_32_S_0_10_10_16_S2|Instruction imm-field relocation|
+`opr1 = pop (), (*(uint32_t *) PC) [9 ... 0] = opr1 [27 ... 18],`
+`(*(uint32_t *) PC) [25 ... 10] = opr1 [17 ... 2]`
+ with check 28-bit signed overflow and 4-bit aligned
+
+|46      |R_LARCH_SOP_POP_32_U              |Instruction fixup|
+`(*(uint32_t *) PC) = pop ()` with check 32-bit unsigned overflow
+
+|47      |R_LARCH_ADD8 |8-bit in-place addition|`*(int8_t *) PC += S + A`
+|48      |R_LARCH_ADD16|16-bit in-place addition|`*(int16_t *) PC += S + A`
+|49      |R_LARCH_ADD24|24-bit in-place addition|`*(int24_t *) PC += S + A`
+|50      |R_LARCH_ADD32|32-bit in-place addition|`*(int32_t *) PC += S + A`
+|51      |R_LARCH_ADD64|64-bit in-place addition|`*(int64_t *) PC += S + A`
+|52      |R_LARCH_SUB8 |8-bit in-place subtraction|`*(int8_t *) PC -= S + A`
```

Comment:
RISC-V style linker relaxation imposes significant complexity on the toolchain and significant costs on object files. If you don't use linker relaxation, ADD/SUB are not needed.

Reply:
> RISC-V style linker relaxation imposes significant complexity on the toolchain and significant costs on object files. If you don't use linker relaxation, ADD/SUB are not needed.

To support such as ".dword sym1 - sym2" hand-writing assemble.